default/AI/ShipDesignAI.py

```diff
@@ -0,0 +1,1602 @@
+- AIShipDesign_Outposter: same but for outposter ships
+- AIShipDesign_Trooper: same but for troop ships
+- AdditionalSpecifications:  Defines all requirements we have for our designs such as minimum fuel or minimum starlane speed.
+
+global variables:
+- Cache: Instance of the ShipDesignCache class - all cached information is stored in here.
+"""
+
+# ToDo: Implement a nicer naming system (based on rating? Hull?)
+# ToDo: add hull.detection to interface, then add scout class
+
+import freeOrionAIInterface as fo
+import FreeOrionAI as foAI
+import ColonisationAI
+import copy
+import traceback
+import math
+from collections import Counter
+from collections import defaultdict
+
+# Define meta classes for the ship parts
+ARMOUR = frozenset({fo.shipPartClass.armour})
+SHIELDS = frozenset({fo.shipPartClass.shields})
+DETECTION = frozenset({fo.shipPartClass.detection})
+STEALTH = frozenset({fo.shipPartClass.stealth})
+FUEL = frozenset({fo.shipPartClass.fuel})
+COLONISATION = frozenset({fo.shipPartClass.colony})
+ENGINES = frozenset({fo.shipPartClass.starlaneSpeed})  # TODO: Update to "speed" after new test build is supplied
+TROOPS = frozenset({fo.shipPartClass.troops})
+WEAPONS = frozenset({fo.shipPartClass.shortRange, fo.shipPartClass.missiles,
+                     fo.shipPartClass.fighters, fo.shipPartClass.pointDefense})
+ALL_META_CLASSES = frozenset({WEAPONS, ARMOUR, DETECTION, FUEL, STEALTH, SHIELDS, COLONISATION, ENGINES, TROOPS})
+
+# Prefixes for the test ship designs
+TESTDESIGN_NAME_BASE = "AI_TESTDESIGN"
+TESTDESIGN_NAME_HULL = TESTDESIGN_NAME_BASE+"_HULL"
+TESTDESIGN_NAME_PART = TESTDESIGN_NAME_BASE+"_PART"
+
+# Hardcoded prefered hullname for testdesigns - should be a hull without conditions but with maximum different slottypes
+TESTDESIGN_PREFERED_HULL = "SH_BASIC_MEDIUM"
+
+
+class ShipDesignCache(object):
+    """This class handles the caching of information used to assess and build shipdesigns in this module.
+
+    Important methods:
+    update_for_new_turn(self): Updates the cach for the current turn, to be called once at the beginning of each turn.
+
+    Important members:
+    testhulls:                 # set of all hullnames used for testdesigns
+    design_id_by_name          # {"designname": designid}
+    part_by_partname           # {"partname": part object}
+    map_reference_design_name  # {"reference_designname": "ingame_designname"}, cf. _build_reference_name()
+    strictly_worse_parts       # strictly worse parts: {"part": ["worsePart1","worsePart2"]}
+    hulls_for_planets          # buildable hulls per planet {planetID: ["buildableHull1","buildableHull2",...]}
+    parts_for_planets          # buildable parts per planet and slot: {planetID: {slottype1: ["part1","part2"]}}
+    best_designs               # {shipclass:{reqTup:{species:{available_parts:{hull:(rating,best_parts)}}}}}
+    production_cost            # {planetID: {"partname1": local_production_cost, "hullname1": local_production_cost}}
```

Comment:
Game are full of features that is not described in game guide. I call them secrets.  Some of this features are waiting its time, some are dead and present as trash piles in code. 

As you notice I comment on `production_cost` because I see strange structure. Looks like this place need same as `loc_invariant = True`